### PR TITLE
Dataset.copy() to preserve encoding

### LINKED ISSUE #1586: Dataset.copy() drops encoding
```
ds = Dataset()
ds.encoding = {"unlimited_dims": 'x'}
ds.copy().encoding
{}
```
By looking at dataset.py, there's a lot of calls to ``Dataset._construct_direct`` that omit the encoding. Is it correct to add it in all cases?

### PR DESCRIPTION
**Pull request by crusaderky**
- [x] Closes #1586
 - [x] Tests added / passed
 - [x] Passes ``git diff upstream/master | flake8 --diff``
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

### CODE REVIEW

**Comment by crusaderky** [comment]
Ready for merge if there's no objections

**Comment by jhamman** [comment]
Pending the outcome of our conversation on #1586, this looks good to me.

File: xarray/tests/test_dataset.py [hunk 1]
```diff
@@ -1434,6 +1435,7 @@ def test_copy(self):
 
         for copied in [data.copy(deep=False), copy(data)]:
             self.assertDatasetIdentical(data, copied)
+            self.assertEqual(data.encoding, copied.encoding)
```
Comment (jhamman) [inline xarray/tests/test_dataset.py]: note on behalf of @MaximilianR -

this can just be `assert data.encoding == copied.encoding`

Comment (crusaderky) [reply xarray/tests/test_dataset.py]: Doesn't assertEqual produce better output in case of failure?

Comment (MaximilianR) [reply xarray/tests/test_dataset.py]: Pytest satisfies all your desires!

Comment (jhamman) [reply xarray/tests/test_dataset.py]: @crusaderky - pytest has nice introspection for this sort of thing. https://docs.pytest.org/en/latest/assert.html#making-use-of-context-sensitive-comparisons

**Pull request merged**